daemon/cmd/kube_proxy_replacement.go

```diff
@@ -361,20 +363,62 @@ func probeKubeProxyReplacementOptions() error {
 	return nil
 }
 
-func probeManagedNeighborSupport() {
+func probeManagedNeighborSupport() error {
 	if option.Config.DryMode {
-		return
+		return nil
+	}
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	oldns, err := netns.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get current netns: %w", err)
+	}
+	defer oldns.Close()
+
+	newns, err := netns.New()
+	if err != nil {
+		return fmt.Errorf("failed to create new netns: %w", err)
+	}
+	defer newns.Close()
+
+	la := netlink.NewLinkAttrs()
+	la.Name = "cilium-dummy"
+
+	dummy := &netlink.Dummy{LinkAttrs: la}
+	if err := netlink.LinkAdd(dummy); err != nil {
+		return fmt.Errorf("failed to add dummy link: %w", err)
 	}
 
-	// Probes for kernel commit:
-	//   856c02dbce4f ("bpf: Introduce helper bpf_get_branch_snapshot")
-	// This is a bit of a workaround given feature probing for netlink
-	// neighboring subsystem is cumbersome. The commit was added in the
-	// same release as managed neighbors, that is, 5.16+.
-	if probes.HaveProgramHelper(ebpf.Kprobe, asm.FnGetBranchSnapshot) == nil {
-		log.Info("Using Managed Neighbor Kernel support")
-		option.Config.ARPPingKernelManaged = true
+	neigh := netlink.Neigh{
+		LinkIndex: dummy.Index,
+		IP:        net.IPv4(10, 1, 1, 1),
+		Flags:     netlink.NTF_EXT_LEARNED,
+		FlagsExt:  netlink.NTF_EXT_MANAGED,
 	}
+
+	if err := netlink.NeighAdd(&neigh); err != nil {
+		return fmt.Errorf("failed to add neighbor: %w", err)
+	}
+
+	nl, err := netlink.NeighList(dummy.Index, 0)
+	if err != nil {
+		return fmt.Errorf("failed to list neighbors: %w", err)
+	}
+
+	for _, n := range nl {
+		if n.IP.Equal(neigh.IP) && n.Flags == netlink.NTF_EXT_LEARNED && n.FlagsExt == netlink.NTF_EXT_MANAGED {
+			log.Info("Using Managed Neighbor Kernel support")
+			option.Config.ARPPingKernelManaged = true
+			break
+		}
+	}
+
+	if err := netns.Set(oldns); err != nil {
+		return fmt.Errorf("failed to switch back to original netns '%s': %w", oldns.UniqueId(), err)
```

Comment:
`defer runtime.UnlockOSThread()` will cause the OS thread to get unlocked even if setting the thread back to the initial netns fails (or if any of the errors above fire, really). This is really bad, as the thread be released back into the schedulable pool, but remain in the empty netns. Same as with bpf objects; `newns` is simply a handle, and the active thread maintains a reference to the netns.

`runtime.UnlockOSThread()` may only be called after a successful call to netns.Set(). If that doesn't succeed, the only safe way to proceed is by letting the OS thread die, which means terminating the current goroutine without calling UnlockOSThread().

In order to do this safely, we'll need to start `probeManagedNeighborSupport` in a goroutine, so we can make sure the goroutine ends when the function exits, giving us control over the lifecycle of the underlying thread.

Additionally, the logic will need to be changed to this:

```go
defer func() {
  // defer closes over named return variable err

  if nerr := netns.Set(oldns); nerr != nil {
    // The current goroutine is locked to an OS thread and we've failed
    // to undo state modifications to the thread. 
    err = fmt.Errorf("error setting thread back to its original netns: %w (original error: %s)", nerr, err)
    return
  }

  runtime.UnlockOSThread()
}
```